File: python/pyfind/pyfind/fileresult.py
# -*- coding: utf-8 -*-
"""
###############################################################################
#
# fileresult.py
#
# class FileResult: encapsulates a file to evaluate as find match and return
#                   if matching
#
###############################################################################
"""
import os.path
from io import StringIO
from pathlib import Path

from .color import ConsoleColor
from .filetypes import FileType
from .findsettings import FindSettings, SortBy


class FileResult:
    """encapsulates a file to evaluate as a find match and return if matching"""
    CONTAINER_SEPARATOR = '!'

    __slots__ = ['containers', 'path', 'file_type', 'file_size', 'last_mod']

    def __init__(self,
                 containers: list[Path] = None,
                 path: Path = None,
                 file_type: FileType = FileType.UNKNOWN,
                 file_size: int = 0,
                 last_mod: float = 0.0):
        self.containers = containers if containers else []
        self.path = path
        self.file_type = file_type
        self.file_size = file_size
        self.last_mod = last_mod

    @property
    def relative_path(self) -> Path:
        """Get relative path of FileResult (does not include any containers)"""
        return self.path

    def __str__(self):
        sio = StringIO()
        if self.containers:
            sio.write(self.CONTAINER_SEPARATOR.join([str(c) for c in self.containers]))
            sio.write(self.CONTAINER_SEPARATOR)
        sio.write(str(self.relative_path))
        return sio.getvalue()

    def __lt__(self, other):
        if self.path.parent == other.path.parent:
            return self.path.name < other.path.name
        return self.path.parent < other.path.parent

    def __eq__(self, other):
        return self.path == other.path


class FileResultFormatter(object):
    """provides formatting of FileResult instances"""

    def __init__(self, settings: FindSettings):
        self.settings = settings
        if settings.colorize:
            if settings.in_dir_patterns:
                self.format_dir_path = self.__format_dir_path_with_color
            if settings.in_extensions or settings.in_file_patterns:
                self.format_file_name = self.__format_file_name_with_color

    @staticmethod
    def colorize(s: str, match_start_index: int, match_end_index: int) -> str:
        """colorize a string"""
        prefix = ''
        if match_start_index > 0:
            prefix = s[0:match_start_index]
        suffix = ''
        if match_end_index < len(s):
            suffix = s[match_end_index:]
        return prefix + \
            ConsoleColor.GREEN + \
            s[match_start_index:match_end_index] + \
            ConsoleColor.RESET + \
            suffix

    def __format_dir_path_with_color(self, dir_path: Path) -> str:
        """format a dir path, highlighting matches with color"""
        formatted_dir = str(dir_path)
        for p in self.settings.in_dir_patterns:
            match = p.search(formatted_dir)
            if match:
                formatted_dir = FileResultFormatter.colorize(
                    formatted_dir, match.start(), match.end())
                break
        return formatted_dir

    def format_dir_path(self, dir_path: Path) -> str:
        """format a dir path, just returns the path as str by default but can be
           redefined to point to __format_dir_path_with_color() if settings
           require colorization"""
        return str(dir_path)

    def __format_file_name_with_color(self, file_name: str) -> str:
        """format a file name, highlighting matches with color"""
        formatted_file_name = file_name
        for p in self.settings.in_file_patterns:
            match = p.search(formatted_file_name)
            if match:
                formatted_file_name = FileResultFormatter.colorize(
                    formatted_file_name, match.start(), match.end())
                break
        if len(self.settings.in_extensions):
            # we know that file_name has a matching extension, no need to match
            idx = formatted_file_name.rfind('.')
            if 0 < idx < len(formatted_file_name) - 1:
                formatted_file_name = FileResultFormatter.colorize(
                    formatted_file_name, idx + 1, len(formatted_file_name))
        return formatted_file_name

    def format_file_name(self, file_name: str) -> str:
        """format a file name, just returns the name by default but can be
           redefined to point to __format_file_name_with_color() if settings
           require colorization"""
        return file_name

    def format_path(self, path: Path) -> str:
        """format a path"""
        parent = '.'
        if path.parent:
            parent = self.format_dir_path(path.parent)
        file_name = self.format_file_name(path.name)
        return os.path.join(parent, file_name)

    def format_file_result(self, result: FileResult) -> str:
        """format a FileResult instance"""
        return self.format_path(result.path)


class FileResultSorter(object):
    """provides sorting of FileResult instances"""

    def __init__(self, settings: FindSettings):
        self.settings = settings

    def _case(self, s: str) -> str:
        if self.settings.sort_case_insensitive:
            return s.casefold()
        return s

    def key_by_file_path(self, r: FileResult) -> list:
        return [[self._case(str(c)) for c in r.containers],
                self._case(str(r.path.parent)),
                self._case(r.path.name)]

    def key_by_file_name(self, r: FileResult) -> list:
        return [self._case(r.path.name),
                [self._case(str(c)) for c in r.containers],
                self._case(str(r.path.parent))]

    def key_by_file_size(self, r: FileResult) -> list:
        return [r.file_size] + self.key_by_file_path(r)

    def key_by_file_type(self, r: FileResult) -> list:
        return [r.file_type] + self.key_by_file_path(r)

    def key_by_last_mod(self, r: FileResult) -> list:
        return [r.last_mod] + self.key_by_file_path(r)

    def get_sort_key_function(self) -> callable:
        """Get the sort key function based on the settings."""
        match self.settings.sort_by:
            case SortBy.FILEPATH:
                return self.key_by_file_path
            case SortBy.FILENAME:
                return self.key_by_file_name
            case SortBy.FILESIZE:
                return self.key_by_file_size
            case SortBy.FILETYPE:
                return self.key_by_file_type
            case SortBy.LASTMOD:
                return self.key_by_last_mod
            case _:
                return self.key_by_file_path

    def sort(self, file_results: list[FileResult]) -> list[FileResult]:
        """Sort the given list of FileResult instances."""
        sort_key_func = self.get_sort_key_function()
        return sorted(file_results, key=sort_key_func,
                      reverse=self.settings.sort_descending)

File: elixir/exfind/lib/fileresult.ex
defmodule ExFind.FileResult do
  @moduledoc """
  Documentation for `ExFind.FileResult`.
  """
  @container_separator "!"

  def container_separator, do: @container_separator

  defstruct containers: [], path: "", name: "", file_type: :unknown, file_size: 0, last_mod: 0

  def new(path, file_name, file_type, file_size, last_mod) do
    __struct__([path: path, name: file_name, file_type: file_type, file_size: file_size, last_mod: last_mod])
  end

  def new(containers, path, file_name, file_type, file_size, last_mod) do
    __struct__([containers: containers, path: path, name: file_name, file_type: file_type, file_size: file_size, last_mod: last_mod])
  end

  def new(args), do: __struct__(args)

  def relative_path(file_result) do
    Path.join(file_result.path, file_result.name)
  end

  def to_string(file_result) do
    container_str = if Enum.empty?(file_result.containers), do: "", else: Enum.join(file_result.containers, @container_separator) <> @container_separator
    container_str <> Path.join(file_result.path, file_result.name)
  end
end

defmodule ExFind.FileResultFormatter do
  @moduledoc """
  Documentation for `ExFind.FileResultFormatter`.
  """

  alias ExFind.StringUtil

  defstruct [:settings, :fn_format_path, :fn_format_file_name]

  def new(settings) do
    fn_format_path = if settings.colorize and !Enum.empty?(settings.in_dir_patterns) do
      fn(settings, path) -> format_path_with_color(settings, path) end
    else
      fn(_settings, path) -> path end
    end
    fn_format_file_name = if settings.colorize and (!Enum.empty?(settings.in_extensions) or !Enum.empty?(settings.in_file_patterns)) do
      fn settings, file_name -> format_file_name_with_color(settings, file_name) end
    else
      fn _settings, file_name -> file_name end
    end
    __struct__([settings: settings, fn_format_path: fn_format_path, fn_format_file_name: fn_format_file_name])
  end

  def colorize(s, start_idx, end_idx) do
    prefix = if start_idx > 0 do
      String.slice(s, 0, start_idx)
    else
      ""
    end
    suffix = if end_idx < String.length(s) do
      String.slice(s, end_idx, String.length(s))
    else
      ""
    end
    match_text = String.slice(s, start_idx, end_idx - start_idx)
    colorized = IO.ANSI.green() <> match_text <> IO.ANSI.reset()
    prefix <> colorized <> suffix
  end

  def format_path_with_color(settings, path) do
    matching_dir_pattern = Enum.find(settings.in_dir_patterns, false, fn p -> Regex.match?(p, path) end)
    if matching_dir_pattern do
      {match_idx, match_length} = Regex.run(matching_dir_pattern, path, return: :index) |> Enum.at(0)
      colorize(path, match_idx, match_idx + match_length)
    else
      path
    end
  end

  def format_path(formatter, path) do
    formatter.fn_format_path.(formatter.settings, path)
  end

  def format_file_name_with_color(settings, file_name) do
    matching_file_pattern = Enum.find(settings.in_file_patterns, false, fn p -> Regex.match?(p, file_name) end)
    formatted_file_name =
      if matching_file_pattern do
        {match_idx, match_length} = Regex.run(matching_file_pattern, file_name, return: :index) |> Enum.at(0)
        colorize(file_name, match_idx, match_idx + match_length)
      else
        file_name
      end
    if Enum.empty?(settings.in_extensions) do
      formatted_file_name
    else
      idx = StringUtil.last_index_of(formatted_file_name, "\.")
      if idx < 1 || idx >= String.length(formatted_file_name) do
        formatted_file_name
      else
        colorize(formatted_file_name, idx + 1, String.length(formatted_file_name))
      end
    end
  end

  def format_file_name(formatter, file_name) do
    formatter.fn_format_file_name.(formatter.settings, file_name)
  end

  def format_file_result(formatter, file_result) do
    parent = formatter.fn_format_path.(formatter.settings, file_result.path)
    file_name = formatter.fn_format_file_name.(formatter.settings, file_result.name)
    Path.join(parent, file_name)
  end
end

defmodule ExFind.FileResultSorter do
  @moduledoc """
  Documentation for `ExFind.FileResultSorter`.
  """

  def get_file_path_mapper(settings) do
    if settings.sort_case_insensitive do
      fn r -> {String.downcase(r.path), String.downcase(r.name)} end
    else
      fn r -> {r.path, r.name} end
    end
  end

  def get_file_name_mapper(settings) do
    if settings.sort_case_insensitive do
      fn r -> {String.downcase(r.name), String.downcase(r.path)} end
    else
      fn r -> {r.name, r.path} end
    end
  end

  def get_file_size_mapper(settings) do
    if settings.sort_case_insensitive do
      fn r -> {r.file_size, String.downcase(r.path), String.downcase(r.name)} end
    else
      fn r -> {r.file_size, r.path, r.name} end
    end
  end

  def get_file_type_mapper(settings) do
    if settings.sort_case_insensitive do
      fn r -> {r.file_type, String.downcase(r.path), String.downcase(r.name)} end
    else
      fn r -> {r.file_type, r.path, r.name} end
    end
  end

  def get_last_mod_mapper(settings) do
    if settings.sort_case_insensitive do
      fn r -> {r.last_mod, String.downcase(r.path), String.downcase(r.name)} end
    else
      fn r -> {r.last_mod, r.path, r.name} end
    end
  end

  def get_file_result_mapper(settings) do
    case settings.sort_by do
      :file_name -> get_file_name_mapper(settings)
      :file_size -> get_file_size_mapper(settings)
      :file_type -> get_file_type_mapper(settings)
      :last_mod  -> get_last_mod_mapper(settings)
      _          -> get_file_path_mapper(settings)
    end
  end

  def sort(settings, results) do
    file_result_mapper = get_file_result_mapper(settings)
    direction = if settings.sort_descending, do: :desc, else: :asc
    Enum.sort_by(results, file_result_mapper, direction)
  end
end
